econml/iv/dml/_dml.py

```diff
@@ -15,78 +15,277 @@
+
+        if self._fit_cate_intercept:
+            add_intercept_trans = FunctionTransformer(add_intercept,
+                                                      validate=True)
+            if featurizer:
+                self._featurizer = Pipeline([('featurize', self._original_featurizer),
+                                             ('add_intercept', add_intercept_trans)])
+            else:
+                self._featurizer = add_intercept_trans
+        else:
+            self._featurizer = self._original_featurizer
+
+    def _combine(self, X, T, fitting=True):
+        if X is not None:
+            if self._featurizer is not None:
+                F = self._featurizer.fit_transform(X) if fitting else self._featurizer.transform(X)
+            else:
+                F = X
+        else:
+            if not self._fit_cate_intercept:
+                raise AttributeError("Cannot have X=None and also not allow for a CATE intercept!")
+            F = np.ones((T.shape[0], 1))
+        return cross_product(F, T)
 
     def fit(self, Y, T, X=None, W=None, Z=None, nuisances=None, sample_weight=None, freq_weight=None, sample_var=None):
         Y_res, T_res, Z_res = nuisances
-        if Z_res.ndim == 1:
-            Z_res = Z_res.reshape(-1, 1)
-        # DMLATEIV is just like 2SLS; first regress T_res on Z_res, then regress Y_res on predicted T_res
-        T_res_pred = self._first_stage.fit(Z_res, T_res,
-                                           sample_weight=sample_weight).predict(Z_res)
-        # TODO: allow the final model to actually use X? Then we'd need to rename the class
-        #       since we would actually be calculating a CATE rather than ATE.
-        self._model_final.fit(X=None, T=T, T_res=T_res_pred, Y_res=Y_res, sample_weight=sample_weight)
+
+        # Track training dimensions to see if Y or T is a vector instead of a 2-dimensional array
+        self._d_t = shape(T_res)[1:]
+        self._d_y = shape(Y_res)[1:]
+
+        XT_res = self._combine(X, T_res)
+        XZ_res = self._combine(X, Z_res)
+        filtered_kwargs = filter_none_kwargs(sample_weight=sample_weight,
+                                             freq_weight=freq_weight, sample_var=sample_var)
+
+        self._model_final.fit(Y_res, XT_res, XZ_res, **filtered_kwargs)
+
         return self
 
     def predict(self, X=None):
-        # TODO: allow the final model to actually use X?
-        return self._model_final.predict(X=None)
+        X2, T = broadcast_unit_treatments(X if X is not None else np.empty((1, 0)),
+                                          self._d_t[0] if self._d_t else 1)
+        XT = self._combine(None if X is None else X2, T, fitting=False)
+        prediction = self._model_final.predict(XT)
+        return reshape_treatmentwise_effects(prediction,
+                                             self._d_t, self._d_y)
 
     def score(self, Y, T, X=None, W=None, Z=None, nuisances=None, sample_weight=None):
```

Comment:
Other potential score could be how much variance of `Y_res - Y_res_pred` can be explained by `Z_res` on the sample. Ideally no variance should be explained. So one could calculate the R2score in this regression and set the score to be 1 - R2score. If R2score=0, then we get the best performance, as no variance is explained and if we get R2score=1, then we very bad. This could be implemented easily using statsmodels: 
```python
1 - sm.OLS(Y_res - Y_res_pred, Z_res).fit().rsquared
```

Reply:
Great catch! Since the score function for other classes is computing MSE,I will take your first suggestion here! To also consider `sample_weight`, I think it should be 
```
np.linalg.norm(np.average(cross_product(Z_res,  Y_res - Y_res_pred) , weights=sample_weight, axis=0), ord=2) 
```